File: src/advent_of_code/y2021/day07.clj
(ns advent-of-code.y2021.day07
  (:require [advent-of-code.support :refer :all]))

(def input (get-input 2021 7))

(defn solve [cost input]
  (let [crabs (find-ints input)]
    (->> (range (apply min crabs) (inc (apply max crabs)))
         (map (fn [pos] (reduce + (map #(cost (abs (- pos %))) crabs))))
         (apply min))))

(def solve1 (partial solve identity))
(def solve2 (partial solve #(/ (* % (inc %)) 2)))

File: src/advent_of_code/y2021/day07.py
IN = sorted(map(int, open("resources/inputs/2021/day07.txt").read().split(",")))
R = range(IN[0], IN[-1] + 1)
print(min(sum(abs(crab - pos) for crab in IN) for pos in R))
print(min(sum(abs(crab - pos) * (abs(crab - pos) + 1) // 2 for crab in IN) for pos in R))
